### PR TITLE
[Feature] Double-signing slashing

### PR DESCRIPTION
**Pull request by Stefan-Ethernal**
# Description

Double signers are malicious validators who have signed more than 1 message of the same type, with different hash in the given round. The aim of this feature is to track such malicious behavior and punish it by slashing a portion of the stake.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: consensus/polybft/checkpoint_manager.go [hunk 1]
```diff
@@ -64,22 +69,36 @@ type checkpointManager struct {
...
 	state *State) *checkpointManager {
-	retry := &eventsGetter[*ExitEvent]{
+	exitEventGetter := &eventsGetter[*ExitEvent]{
 		blockchain: blockchain,
 		isValidLogFn: func(l *types.Log) bool {
 			return l.Address == contracts.L2StateSenderContract
 		},
 		parseEventFn: parseExitEvent,
 	}
+	slashedEventGetter := &eventsGetter[*contractsapi.SlashedEvent]{
```
Comment (goran-ethernal) [inline consensus/polybft/checkpoint_manager.go]: You can use the same event getter for both events. just set it up like `eventsGetter[contractsapi.EventAbi]`, then in  `isValidLogFn` plugin function, check if log address is from ValidatorSet or L2StateSender contracts, and on the place where you call event getter just use switch case by type to separate them into two arrays. That way we don't need to go through the same blocks to get two types of events.

Small optimization per say :smile: .

File: consensus/polybft/consensus_runtime.go [hunk 2]
```diff
@@ -167,6 +173,13 @@ func newConsensusRuntime(log hcf.Logger, config *runtimeConfig) (*consensusRunti
 		return nil, fmt.Errorf("consensus runtime creation - restart epoch failed: %w", err)
 	}
 
+	tracker, err := slashing.NewDoubleSigningTracker(logger.Named("double_sign_tracker"), config.State.StakeStore)
```
Comment (goran-ethernal) [inline consensus/polybft/consensus_runtime.go]: Can we move this to a separate function to be consistent to other init functions like `initGovernanceManager`, `initStakeManager`, etc.

File: consensus/polybft/consensus_runtime.go [hunk 3]
```diff
@@ -333,22 +346,37 @@ func (c *consensusRuntime) OnBlockInserted(fullBlock *types.FullBlock) {
...
+	if err := c.doubleSigningTracker.PostBlock(postBlock); err != nil {
+		c.logger.Error("post block callback failed in double signing tracker", "err", err)
+	}
+
+	// handle governance events that happened in block
+	if err := c.governanceManager.PostBlock(postBlock); err != nil {
+		c.logger.Error("failed to post block in governance manager", "err", err)
+	}
+
+	// handle governance events that happened in block
```
Comment (goran-ethernal) [inline consensus/polybft/consensus_runtime.go:389]: Is this `c.governanceManager.PostBlock(postBlock)` mistakenly pasted twice?

Comment (Stefan-Ethernal) [reply consensus/polybft/consensus_runtime.go:389]: Tnx good point, merge conflicts...

Comment (Stefan-Ethernal) [reply consensus/polybft/consensus_runtime.go:389]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1808/commits/437e00809243655bfaa6e1d26859197d6c407c42

File: consensus/polybft/consensus_runtime.go (hunk 2 above)
Comment (Stefan-Ethernal) [reply consensus/polybft/consensus_runtime.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1808/commits/437e00809243655bfaa6e1d26859197d6c407c42

**Review by Nemanja0x — APPROVED** [review]

**Review by igorcrevar — APPROVED** [review]

**Review by goran-ethernal — APPROVED** [review]

File: consensus/polybft/checkpoint_manager.go (hunk 1 above)
Comment (Stefan-Ethernal) [reply consensus/polybft/checkpoint_manager.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1808/commits/14b5cfb373f1057b108c0ed83c0f4ffe4e369289

**Pull request merged**